packages/neo-one-server-client/src/npmCheck.ts

```diff
@@ -0,0 +1,90 @@
+import { getEscapedNPMName } from '@neo-one/utils';
+import fetch from 'cross-fetch';
+import * as fs from 'fs-extra';
+
+export interface NPMResult {
+  readonly success: boolean;
+  readonly version: string;
+  readonly date: number;
+}
+const emptyNPMResult: NPMResult = {
+  success: false,
+  version: '',
+  date: 0,
+};
+
+const askNPM = async (pkg: string): Promise<NPMResult> =>
+  new Promise<NPMResult>(async (resolve) => {
+    const NPM_REGISTRY_URL = 'https://registry.yarnpkg.com';
+    const escapedName = getEscapedNPMName(pkg);
+    const url = `${NPM_REGISTRY_URL}/${escapedName}`;
+
+    await fetch(url).then(async (res) => {
+      if (res.ok) {
+        try {
+          const json = await res.json();
+          const version = json['dist-tags'].latest;
+          const npmResult: NPMResult = {
+            success: true,
+            version,
+            date: Date.parse(json.time[version]),
+          };
+          resolve(npmResult);
+        } catch {
+          // ignore errors...
+        }
+      }
+      resolve(emptyNPMResult);
+    });
+  });
+
+const promiseTimeout = async <t>(delay: number): Promise<t> =>
+  new Promise<t>((resolve) => {
+    setTimeout(resolve, delay, {});
+  });
+
+const saveResult = async (file: string, result: NPMResult) => {
+  await fs.writeFile(file, JSON.stringify(result));
+};
+
+const triggerCacheUpdate = async (pkg: string, cachePath: string, timeout: number): Promise<void> => {
+  const result: NPMResult = await Promise.race([askNPM(pkg), promiseTimeout<NPMResult>(timeout)]);
+  await saveResult(cachePath, {
+    version: result.version,
+    date: Date.now(),
+    success: true,
+  });
+};
+
+const readCache = async (file: string) => {
+  try {
+    const contents = await fs.readFile(file, 'utf8');
+    if (contents.length > 0) {
+      return JSON.parse(contents);
+    }
+  } catch {
+    // ignore errors
+  }
+
+  return undefined;
+};
+
+export const npmCheck = async (
+  pkg: string,
+  cachePath: string,
+  timeout: number,
+  checkDelaySec: number,
+): Promise<NPMResult> => {
+  const message = await readCache(cachePath);
+
+  if (message === undefined) {
```

Comment:
Doesn't seem like the cache will ever be filled with this check. I think you can remove this.

Reply:
After the first run, it should be.  It's never deleted.

Reply:
Right, but if there is no file, we return an `emptyNPMResult`, and we never trigger a write to the file.

Reply:
Thank you.